Normative: Provide source text to HostEnsureCanCompileStrings

ptomato opened the pull request:
This is based on Mike Samuel's work in #1498, but ultimately goes in a different direction. Due to #2670, this change can be made simpler than it previously was.

This change provides the source text to be evaluated, and the grammar symbol that should be used to parse it, to the host hook HostEnsureCanCompileStrings.

One example of where this is needed is for allowing a Content Security Policy to provide hashes for code executed via `eval()` or `new Function()`:
https://github.com/w3c/webappsec-csp/issues/623
This is useful on its own, but has come up again in the topic of ShadowRealm-HTML integration. In a ShadowRealm you can either execute code asynchronously, with ShadowRealm.p.importValue, or synchronously, with ShadowRealm.p.evaluate. Because the latter uses `eval()` inside the ShadowRealm, it's subject to CSP rules, so the only CSP policy that will let you execute synchronously in the realm is `unsafe-eval`.

~~The original purpose of #1498 was to support Trusted Types, which is still a goal of this PR.~~

This is a separate needs-consensus PR, rather than being part of the ShadowRealm proposal, because it's useful independently of ShadowRealm, and also ShadowRealm would go forward regardless of whether this goes forward.

Prior art: https://github.com/tc39/proposal-dynamic-code-brand-checks

<!--
If you are changing the signature or behavior of an existing construct, please check if this affects downstream dependencies (searching for the construct's name is sufficient) and if needed file an issue:

* [Web IDL](https://webidl.spec.whatwg.org/) — [file an issue](https://github.com/whatwg/webidl/issues/new)
* [HTML Standard](https://html.spec.whatwg.org/) — [file an issue](https://github.com/whatwg/html/issues/new)
* [ECMAScript Intl API](https://tc39.es/ecma402/) - [file an issue](https://github.com/tc39/ecma402/issues/new)

Note: please ensure that the "Allow edits and access to secrets by maintainers" checkbox remains checked.
-->

ptomato commented:
cc @mikesamuel

ljharb commented on spec.html:
```diff
@@ -28965,13 +28966,27 @@ <h1>
+
+        <p>
+          Implementations of HostEnsureCanCompileStrings should not coerce _bodyText_ to a string since callers will later do that and HostEnsureCanCompileStrings must not base security decisions on both coercions producing the same string.
+          Clients may base security decisions on internal slots.
```
what does this mean?

ptomato replied:
This note I included as it was in #1498 so maybe @mikesamuel can correct me if I'm wrong. My understanding is that if you try to compile an object as code (which cannot happen through `eval(myObject)`, but can happen through `new Function('param', myObject)` it's valid for the host to use `myObject`'s internal slots to decide whether to block this operation or not.

jmdyck replied:
```suggestion
          If _bodyText_ is an Object, the host may examine its internal slots.
```

ljharb replied:
Does that need to be specified? Can't hosts make that decision based on any criteria they want?

ptomato replied:
Yeah, I guess it can be removed, as it's implicitly allowed.

I assume this line was originally intended to emphasize that if _bodyText_ is a [`TrustedScript`](https://w3c.github.io/trusted-types/dist/spec/#trusted-script), the host should not call its `toString()` method but instead examine its [[Data]] internal slot.

ljharb commented on spec.html:
```diff
@@ -28965,13 +28966,27 @@ <h1>
+        </emu-note>
+
+        <p>
+          Implementations of HostEnsureCanCompileStrings should not coerce _bodyText_ to a string since callers will later do that and HostEnsureCanCompileStrings must not base security decisions on both coercions producing the same string.
+          Clients may base security decisions on internal slots.
```
these kind of feel like they should be a list instead of a p?

jmdyck commented on spec.html:
```diff
@@ -28965,13 +28966,27 @@ <h1>
+
+        <emu-note>
+          <p>
+            _goal_ is the production which will be used to parse _bodyText_.
```
```suggestion
            _goal_ is the nonterminal that will be used to parse _bodyText_.
```

jmdyck commented on spec.html:
```diff
@@ -28965,13 +28966,27 @@ <h1>
+        </emu-note>
+
+        <p>
+          Implementations of HostEnsureCanCompileStrings should not coerce _bodyText_ to a string since callers will later do that and HostEnsureCanCompileStrings must not base security decisions on both coercions producing the same string.
```
> should not coerce `_bodyText_` to a string

According to the header, `_bodyText_` already *is* a String.

jmdyck commented on spec.html:
```diff
@@ -30185,6 +30198,8 @@ <h1>
               1. Let _parameterSym_ be the grammar symbol |FormalParameters[+Yield, +Await]|.
               1. Let _fallbackProto_ be *"%AsyncGeneratorFunction.prototype%"*.
+            1. Let _currentRealm_ be the current Realm Record.
+            1. Perform ? HostEnsureCanCompileStrings(_currentRealm_, _bodySym_, _bodyArg_).
```
`_bodyArg_` is an ECMAScript language value, but HostEnsureCanCompileStrings says it expects a String.

jmdyck commented on spec.html:
```diff
@@ -19219,7 +19219,7 @@ <h1>Runtime Semantics: Evaluation</h1>
               1. Let _evalArg_ be the first element of _argList_.
               1. If the source text matched by this |CallExpression| is strict mode code, let _strictCaller_ be *true*. Otherwise let _strictCaller_ be *false*.
-              1. [id="step-callexpression-evaluation-direct-eval"] Return ? PerformEval(_evalArg_, _strictCaller_, *true*).
+              1. [id="step-callexpression-evaluation-direct-eval"] Return ? PerformEval(_evalArg_, |Script|, _strictCaller_, *true*).
```
```suggestion
              1. [id="step-callexpression-evaluation-direct-eval"] Return ? PerformEval(_evalArg_, _strictCaller_, *true*).
```

There's no need to pass `|Script|` to PerformEval; PerformEval is 'hard-coded' to use `|Script|`.

ptomato replied:
Good point, that makes this PR even smaller.

jmdyck commented on spec.html:
```diff
@@ -28884,13 +28884,14 @@ <h1>eval ( _x_ )</h1>
       <p>It performs the following steps when called:</p>
       <emu-alg>
-        1. Return ? PerformEval(_x_, *false*, *false*).
+        1. Return ? PerformEval(_x_, |Script|, *false*, *false*).
```
```suggestion
        1. Return ? PerformEval(_x_, *false*, *false*).
```

jmdyck commented on spec.html:
```diff
@@ -28884,13 +28884,14 @@ <h1>eval ( _x_ )</h1>
         <h1>
           PerformEval (
             _x_: an ECMAScript language value,
+            _goal_: a grammar symbol,
```
```suggestion
```

jmdyck commented on spec.html:
```diff
@@ -28902,7 +28903,7 @@ <h1>
           1. Let _evalRealm_ be the current Realm Record.
           1. NOTE: In the case of a direct eval, _evalRealm_ is the realm of both the caller of `eval` and of the `eval` function itself.
-          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_).
+          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_, _goal_, _x_).
```
```suggestion
          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_, |Script|, _x_).
```

jmdyck commented on spec.html:
```diff
@@ -28965,13 +28965,27 @@ <h1>
+            _goal_ is the nonterminal which will be used to parse _bodyText_.
+            For example, if called via %Function% it might be the grammar symbol |FunctionBody[~Yield, ~Await]| and from %eval% it might be the grammar symbol |Script|.
+          </p>
+        </emu-note>
```
Since the Note is about a parameter, it should maybe go at the top, so that it appears after the para that introduces the parameters.

ptomato replied:
Agreed. I guess it probably doesn't need to be a Note, either, since it's just a description of the parameter.

jmdyck commented on spec.html:
```diff
@@ -28965,13 +28965,27 @@ <h1>
+          </p>
+        </emu-note>
+
+        <ul>
```
Looking at the rendering, I think these should be two plain paragraphs rather than a list.

Also, they should maybe go before the "default implementation" para, as that's usually the last thing that a host-hook section talks about.

jmdyck commented on spec.html:
```diff
@@ -30185,6 +30197,8 @@ <h1>
               1. Let _parameterSym_ be the grammar symbol |FormalParameters[+Yield, +Await]|.
               1. Let _fallbackProto_ be *"%AsyncGeneratorFunction.prototype%"*.
+            1. Let _currentRealm_ be the current Realm Record.
+            1. Perform ? HostEnsureCanCompileStrings(_currentRealm_, _bodySym_, _bodyArg_).
```
I'm curious why `_bodyArg_` is submitted for approval but not `_parameterArgs_`. A FormalParameter can include an Initializer; doesn't that make it as 'suspect' as a FunctionBody?

Moreover, is it possible that there are security policies that can't be enforced unless you have both the body *and* the parameters? That is, I'm wondering if the HostEnsureCanCompileStrings call should be delayed until you can pass `_sourceString_` to it.

ptomato replied:
Tricky. You are absolutely right about _parameterArgs_ potentially being executable code. This maybe needs some more thinking about what the developer experience should look like in regard to https://github.com/w3c/webappsec-csp/issues/623. It seems pretty bad if the bundling tool has to calculate its own version of _sourceString_ in order to hash that.

If we did move the call later and pass _sourceString_, we'd also still have to pass _bodyArg_, otherwise the host wouldn't be able to distinguish when the code came from a `TrustedScript`.

ptomato replied:
So, one possibility is that we leave out Trusted Types as a use case for the time being, to be accommodated with a future normative PR. (If this PR was going to accommodate Trusted Types correctly, it'd have to remove step 2 of [PerformEval](https://tc39.es/ecma262/#sec-performeval) anyway, I realized.)

Another possibility being discussed on the CSP side is that `unsafe-eval-hashes` doesn't necessarily have to allow the multi-argument `new Function()` at all, so if no other host was going to make a distinction based on the hashes of the parameter args, we could get away with passing a boolean _hasParameterArgs_, and hosts with CSP could just choose to deny any source string for which that argument was true.

ljharb replied:
The former seems best given that trusted types hasn’t been discussed in awhile nor enthusiastically received in plenary.

michaelficarra commented on spec.html:
```diff
@@ -28965,12 +28965,19 @@ <h1>
         <h1>
           HostEnsureCanCompileStrings (
             _calleeRealm_: a Realm Record,
+            _goal_: a grammar symbol,
```
If only certain nonterminal will be passed, I would prefer to enumerate exactly those. Otherwise, I would prefer to take an enum.

ptomato replied:
Changed to a list of possible nonterminals.

ptomato commented:
Updated to reflect the consensus in TC39 plenary of 2023-11-27:

- `eval()` and friends pass their argument to HostEnsureCanCompileStrings
- `new Function()` and friends pass their unstringified argument to HostEnsureCanCompileStrings, only if there is a single argument. If there is more than one argument, the previous behaviour is maintained.
- In the case of direct eval, we inform HostEnsureCanCompileStrings of this.

caridy commented:
thanks for the update @ptomato, this is looking good!

nicolo-ribaudo commented on spec.html:
```diff
@@ -28965,12 +28965,18 @@ <h1>
           HostEnsureCanCompileStrings (
             _calleeRealm_: a Realm Record,
+            optional _direct_: a Boolean,
+            optional _bodyText_: an ECMAScript language value,
```
Very minor nit, but I'd swap the order of these two parameters given that the source text is "more important"

nicolo-ribaudo commented on spec.html:
```diff
@@ -28965,12 +28965,18 @@ <h1>
         </dl>
+        <p>
+          If _bodyText_ is present, or _direct_ signifying whether the evaluation is a direct eval, the host environment may base its decision on their values.
+          Implementations of HostEnsureCanCompileStrings should not coerce _bodyText_ to a string, since callers will do that later, and HostEnsureCanCompileStrings must not base security decisions on both coercions producing the same string.
```
```suggestion
```

`_bodyText_` is already a string.

jmdyck replied:
No, e.g. if you call `Function(42)`, then `_bodyText_` is a Number, not a String.

(`_bodyText_` should probably be renamed.)

ptomato commented:
I've updated this to reflect the option that had the most support in today's plenary meeting: passing the body string to the host hook after stringifying it.

(I've also removed the references to #1498 from the commit message and reset the author; this PR no longer subsumes the goal of #1498, and in fact has ship-of-Theseus'ed so that it now contains nothing from the original)

cc @erights for further discussion

michaelficarra commented on spec.html:
```diff
@@ -28965,12 +28965,17 @@ <h1>
           <dd>It allows host environments to block certain ECMAScript functions which allow developers to interpret and evaluate strings as ECMAScript code.</dd>
         </dl>
+        <p>
+          If _bodyText_ is present, or _direct_ signifying whether the evaluation is a direct eval, the host environment may base its decision on their values.
```
This sentence has some sort of grammatical error.

michaelficarra commented on spec.html:
```diff
@@ -28902,7 +28902,7 @@ <h1>
           1. Let _evalRealm_ be the current Realm Record.
           1. NOTE: In the case of a direct eval, _evalRealm_ is the realm of both the caller of `eval` and of the `eval` function itself.
-          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_).
+          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_, _direct_, _x_).
```
These arguments are in the wrong order.

michaelficarra commented on spec.html:
```diff
@@ -28965,12 +28965,18 @@ <h1>
         </dl>
+        <p>
+          _direct_ signifies whether the evaluation is a direct eval.
+          If _bodyText_ or _direct_ are present, the host environment may base its decision on their values.
```
Without a specific prohibition, this is already true. Is it necessary to state this explicitly?

ptomato replied:
Fair enough.

michaelficarra commented on spec.html:
```diff
@@ -28965,14 +28965,16 @@ <h1>
           <dt>description</dt>
           <dd>It allows host environments to block certain ECMAScript functions which allow developers to interpret and evaluate strings as ECMAScript code.</dd>
         </dl>
+        <p>_paramsStrings_ represents the strings that, when using the `Function` constructor, will be concatenated tgether to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
```
```suggestion
        <p>_paramsStrings_ represents the strings that, when using the `Function` constructor, will be concatenated together to build the parameter list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
```

michaelficarra replied:
You say `_paramsStrings_` here but it's `_parametersStrings_` below.

ptomato commented:
This is now updated with @nicolo-ribaudo's changes based on a conversation with Mark.

bakkot commented on spec.html:
```diff
@@ -29024,12 +29024,17 @@ <h1>
           <dt>description</dt>
           <dd>It allows host environments to block certain ECMAScript functions which allow developers to interpret and evaluate strings as ECMAScript code.</dd>
         </dl>
+        <p>_paramsStrings_ represents the strings that, when using the `Function` constructor, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
```
```suggestion
        <p>_paramsStrings_ represents the strings that, when using one of the `Function`, `GeneratorFunction`, `AsyncFunction`, or `AsyncGeneratorFunction` constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
```

or some other wording - it's not just `Function`, is the point.

nicolo-ribaudo replied:
```suggestion
        <p>_paramsStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
```

if we want to have it a more readable (even though less explicit) sentence

jmdyck commented on spec.html:
```diff
@@ -30263,18 +30266,21 @@ <h1>
               1. Let _parameterSym_ be the grammar symbol |FormalParameters[+Yield, +Await]|.
               1. Let _fallbackProto_ be *"%AsyncGeneratorFunction.prototype%"*.
             1. Let _argCount_ be the number of elements in _parameterArgs_.
+            1. Let _parametersStrings_ be a new empty List.
```
Pluralization is inconsistent between _parameterArgs_ and _parametersStrings_.
```suggestion
            1. Let _parameterStrings_ be a new empty List.
```
(And similarly in the HostEnsureCanCompileStrings parameter.)

jmdyck commented on spec.html:
```diff
@@ -30263,18 +30266,21 @@ <h1>
                 1. Set _k_ to _k_ + 1.
-            1. Let _bodyString_ be the string-concatenation of 0x000A (LINE FEED), ? ToString(_bodyArg_), and 0x000A (LINE FEED).
-            1. Let _sourceString_ be the string-concatenation of _prefix_, *" anonymous("*, _P_, 0x000A (LINE FEED), *") {"*, _bodyString_, and *"}"*.
+            1. Let _bodyString_ be ? ToString(_bodyArg_).
```
Note that this changes the String used to define _body_. I don't *think* it matters, but I'm not sure with this AO.

nicolo-ribaudo replied:
_sourceString_ is still the same anyway, right?

ptomato replied:
It seems a bit odd to me, that the concatenated surrounding `\n`s were passed to ParseText in the first place, but I don't think it makes a difference. I haven't checked, but I'd guess that engines did it this way already, otherwise line numbers in resulting SyntaxErrors would be off by one.

jmdyck replied:
Re the inserted newlines:

They first appeared, in _sourceText_, when that was introduced in PR #697.

Then PR #1479 modified things so that they *also* appeared in _bodyText_. And that's the only substantive change that that PR made, so this change would undo that PR.

I don't entirely understand the issue, but it seems to have something to do with HTML comments and web reality.

nicolo-ribaudo replied:
Oh yes, https://tc39.es/ecma262/#prod-annexB-SingleLineHTMLCloseComment requires a leading line terminator.

Then the order to preserve the correct behavior should be:
- call ToString(...)
- pass the result to the host hook
- add the `\n`s before and after
- pass it to ParseText(StringToCodePoints(_bodyString_), _bodySym_)

jmdyck replied:
Yeah, if the host hook doesn't want the newline-encrusted string.

ptomato replied:
I've pushed a new update that should address this discrepancy. We now make a distinction between _bodyString_ (without added newlines) and _bodyParseString_ (with added newlines).

michaelficarra commented on spec.html:
```diff
@@ -29024,12 +29024,17 @@ <h1>
           <dd>It allows host environments to block certain ECMAScript functions which allow developers to interpret and evaluate strings as ECMAScript code.</dd>
         </dl>
+        <p>_parameterStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.</p>
+        <p>_direct_ signifies whether the evaluation is a direct eval.</p>
```
I would combine this with the previous paragraph, but this is also fine.

bakkot commented on spec.html:
```diff
@@ -30263,22 +30266,26 @@ <h1>
+            1. Let _bodyString_ be ? ToString(_bodyArg_).
+            1. Let _currentRealm_ be the current Realm Record.
+            1. Perform ? HostEnsureCanCompileStrings(_currentRealm_, _parameterStrings_, _bodyString_, *false*).
+            1. Let _bodyParseString_ be the string-concatenation of 0x000A (LINE FEED), _bodyString_, and 0x000A (LINE FEED).
```
It's a little weird to construct this string only after the call to `HostEnsureCanCompileStrings`, but construct `_P_` before it. I'm with either doing both before or both after, with a mild preference for doing both after.

jmdyck replied:
Yeah, I think doing both after is a little easier to understand.

ptomato commented:
@tc39/ecma262-editors I realized this one wasn't in the list of things that would be added before ES2024 is cut. But I believe all of the discussions have been resolved, is there anything else preventing it from making the cut?
